infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/PVCProvisioner.java

```diff
@@ -0,0 +1,214 @@
+import org.eclipse.che.workspace.infrastructure.kubernetes.Names;
+import org.eclipse.che.workspace.infrastructure.kubernetes.environment.KubernetesEnvironment;
+import org.eclipse.che.workspace.infrastructure.kubernetes.environment.KubernetesEnvironment.PodData;
+
+/**
+ * Helps to work with {@link PersistentVolumeClaim} and provision them to {@link
+ * KubernetesEnvironment}.
+ *
+ * @author Sergii Leshchenko
+ */
+public class PVCProvisioner {
+
+  private final String pvcNamePrefix;
+  private final String pvcAccessMode;
+  private final String pvcQuantity;
+  private final PodsVolumes podsVolumes;
+
+  @Inject
+  public PVCProvisioner(
+      @Named("che.infra.kubernetes.pvc.name") String pvcNamePrefix,
+      @Named("che.infra.kubernetes.pvc.quantity") String pvcQuantity,
+      @Named("che.infra.kubernetes.pvc.access_mode") String pvcAccessMode,
+      PodsVolumes podsVolumes) {
+    this.pvcAccessMode = pvcAccessMode;
+    this.pvcQuantity = pvcQuantity;
+    this.pvcNamePrefix = pvcNamePrefix;
+    this.podsVolumes = podsVolumes;
+  }
+
+  /**
+   * Converts {@link Volume} specified in {@link MachineConfig#getVolumes()} to {@link
+   * PersistentVolumeClaim}s and provision them to {@link KubernetesEnvironment}. The machines
+   * corresponding pods and containers are updated in accordance.
+   *
+   * @param k8sEnv environment to provision
+   * @param workspaceId identifier of workspace to which the specified environment belongs to
+   */
+  public void convertCheVolumes(KubernetesEnvironment k8sEnv, String workspaceId) {
+    Map<String, PersistentVolumeClaim> volumeName2PVC =
+        groupByVolumeName(k8sEnv.getPersistentVolumeClaims().values());
+
+    for (PodData pod : k8sEnv.getPodsData().values()) {
+      final PodSpec podSpec = pod.getSpec();
+      List<Container> containers = new ArrayList<>();
+      containers.addAll(podSpec.getContainers());
+      containers.addAll(podSpec.getInitContainers());
+      for (Container container : containers) {
+        final String machineName = Names.machineName(pod, container);
+        InternalMachineConfig machineConfig = k8sEnv.getMachines().get(machineName);
+        if (machineConfig == null) {
+          continue;
+        }
+        Map<String, Volume> volumes = machineConfig.getVolumes();
+        addMachineVolumes(workspaceId, k8sEnv, volumeName2PVC, pod, container, volumes);
+      }
+    }
+  }
+
+  /**
+   * Provision the specified PVCs to the environment.
+   *
+   * <p>Note that:<br>
+   * - PVC is not provisioned if environment already contains PVC for corresponding volume;<br>
+   * - PVC is provisioned with generated unique name;<br>
+   * - corresponding PVC references in Kubernetes Environment are updated during provisioning;<br>
+   *
+   * @param k8sEnv environment to provision
+   * @param toProvision PVCs that should be provisioned to the environment
+   */
+  public void provision(
```

Comment:
Consider moving this method back to `UniqueWorkspacePVCStrategy` as it looks to only be used there. It's confusing to read here, as `#convertCheVolumes()` is used in the common strategy.

Reply:
I wanted to make strategies maximum small, declarative and clear.
`PVCProvisioner` is a class that holds logic for provisioning new PVCs to existing Kubernetes Environment(does not matter based on Che Volumes or provided PVCs list).
I think In this way, it is easier to understand/test whole behavior and pieces with implementation details.
@amisevsk Does it make sense for you? Maybe you have an idea how to improve PVC Strategies not to keep all stuff in one place.

Reply:
That works for me.